Change go version detection logic

soltysh opened the pull request:
Since some of us are already using go 1.7 we're missing certain verification mechanism out of the box. I've changed the logic to check only minimal go version. @stevekuznetsov ptal this part

Unfortunately, govet still complains about:
```
$ hack/verify-govet.sh 
pkg/auth/ldaputil/client.go:69: assignment copies lock value to c: crypto/tls.Config contains sync.Once contains sync.Mutex
pkg/cmd/util/net.go:126: assignment copies lock value to transport: net/http.Transport contains sync.Mutex
```
@deads2k @liggitt any ideas how to fix those?

stevekuznetsov commented on hack/lib/util/golang.sh:
```diff
@@ -0,0 +1,36 @@
+# This library holds golang related utility functions.
+
+# Ensure the go tool exists and is a viable version.
+os::golang::verify_go_version() {
```
we want to use the following func formatting:
```sh
function name() {

}
readonly -f name
```

stevekuznetsov replied:
To explain: the formatting is for consistency, the readonly is for scripts that call other scripts where both `source hack/lib/init.sh`

stevekuznetsov commented on hack/lib/util/golang.sh:
```diff
@@ -0,0 +1,36 @@
+#
+# This library holds golang related utility functions.
+
+# Ensure the go tool exists and is a viable version.
```
we use Godoc-style semantics for comments:
```sh
# os::golang::verify_version ensures that a viable version of Golang is installed
function os::golang::verify_version() {

}
```

stevekuznetsov commented on hack/lib/util/golang.sh:
```diff
@@ -0,0 +1,36 @@
+
+# Ensure the go tool exists and is a viable version.
+os::golang::verify_go_version() {
+  if [[ -z "$(which go)" ]]; then
```
It is better to check the return code, not the output:
```sh
if ! which go >/dev/null 2>&1; then
    # can't find go
fi
```

stevekuznetsov commented on hack/lib/util/golang.sh:
```diff
@@ -0,0 +1,36 @@
+# Ensure the go tool exists and is a viable version.
+os::golang::verify_go_version() {
+  if [[ -z "$(which go)" ]]; then
+    cat <<EOF
```
We should be trying to move to use:
```sh
os::log::info <<EOF whatever
you
want
EOF
```

soltysh replied:
Yeah, I couldn't find anything appropriate, etoomuch bash ;)

stevekuznetsov replied:
Yuck, the current implementation of `os::log::info` is not the best and it doesn't honor newlines, so I think you'll have to do:
```sh
os::log::info "first line"
os::log::info "second line"
os::log::info "third line"
```

That actually might look better too since it will all be formatted the same and line up and stuff

stevekuznetsov commented on hack/lib/util/golang.sh:
```diff
@@ -0,0 +1,36 @@
+
+  local go_version
+  go_version=($(go version))
+  if [[ "${go_version[2]}" < "go1.6" && "${go_version[2]}" != "devel" ]]; then
```
I'm not sure this check will be sufficient -- we don't currently support `go1.7`, and on the test machines we *don't* want to be running any of these steps with `go1.7` until we choose to, since otherwise an overnight `go` install could break all of our jobs. I think what we could do is provide an override flag for users like you who are doing things locally, but the argument before when I proposed that was we don't want people locally seeing everything pass then going to Jenkins and seeing failures, since we don't have a guarantee of backwards compat for stuff like `gofmt` or `govet`.

soltysh replied:
Ok, I'll rework it in such a way that it will always fail for != go1.6, not like currently if go1.6 is not there we'll just skip this over and you never see a failure. But I'll provide a hint in the error message how to bypass the problem.

stevekuznetsov replied:
Whatever we end up using for the flag, we should have it allow *any* golang version and when doing the version check it should print out a `os::log::warning` that we found a golang version that isn't optimal, but we're going on anyway, and there may be some side-effects

stevekuznetsov commented on hack/lib/util/golang.sh:
```diff
@@ -0,0 +1,36 @@
+}
+
+# Ensure the golint tool exists.
+os::golang::verify_golint_version() {
```
Same Bash comments apply to this function

liggitt commented:
> @deads2k @liggitt any ideas how to fix those?

for the first, not really... technically copying the config *is* wrong because the new one shares a map of session tickets, guarded by a copy of a mutex, but that's only used by servers, and this is for a client. also, go makes the same copy.

for the second, we should be using `net.SetTransportDefaults` rather than making a copy of the default client's transport

soltysh commented:
With `PERMISSIVE_GO` these `govet` errors won't show-up in travis, but I'll open separate issues to address them.

@stevekuznetsov comments addressed, ptal once again.

stevekuznetsov commented on hack/lib/util/golang.sh:
```diff
@@ -0,0 +1,32 @@
+# os::golang::verify_go_version ensure the go tool exists and is a viable version.
+function os::golang::verify_go_version() {
+  if ! which go &>/dev/null; then
+    os::log::info "Can't find 'go' in PATH, please fix and retry."
```
This was my fault in the previous comment, but these should be `os::log::error` since they mean something failed.

stevekuznetsov commented on hack/lib/util/golang.sh:
```diff
@@ -0,0 +1,32 @@
+
+# os::golang::verify_go_version ensure the go tool exists and is a viable version.
+function os::golang::verify_go_version() {
+  if ! which go &>/dev/null; then
```
use a tab instead of two spaces to indent

soltysh replied:
You should make your mind, because all over our codebase we use 2 space, 4 spaces and tab interchangeably ;)

stevekuznetsov replied:
*my* scripts should be using a tab... the old `hack/util.sh` crap uses a different thing every ten lines ...

soltysh commented:
@stevekuznetsov updated

stevekuznetsov commented on hack/lib/util/golang.sh:
```diff
@@ -0,0 +1,32 @@
+    os::log::info "Origin requires go version 1.6."
+    os::log::info "Please install Go version 1.6 or use PERMISSIVE_GO=y to bypass this check."
+    return 1
+  fi
```
I would make this look like:
```sh
if [[ "${go_version[2]}" != "go1.6" ]]; then
	os::log::info "Detected go version: ${go_version[*]}."
	if [[ -z "${PERMISSIVE_GO:-}" ]]; then
		os::log::error "Please install Go version 1.6 or use PERMISSIVE_GO=y to bypass this check."
		return 1
	else
		os::log::warn "Detected golang version doesn't match preferred Go version for Origin."
		os::log::warn "This version mismatch could lead to differences in execution between this run and the Origin CI systems."
		return 0
	fi
fi
```

stevekuznetsov commented on hack/lib/util/golang.sh:
```diff
@@ -0,0 +1,32 @@
+	if [[ "${go_version[2]}" != "go1.6" && -z "${PERMISSIVE_GO:-}" ]]; then
+		os::log::error "Detected go version: ${go_version[*]}."
+		os::log::error "Origin requires go version 1.6."
+		os::log::error "Please install Go version 1.6 or use PERMISSIVE_GO=y to bypass this check."
```
I would make this look like:
```sh
if [[ "${go_version[2]}" != "go1.6" ]]; then
	os::log::info "Detected go version: ${go_version[*]}."
	if [[ -z "${PERMISSIVE_GO:-}" ]]; then
		os::log::error "Please install Go version 1.6 or use PERMISSIVE_GO=y to bypass this check."
		return 1
	else
		os::log::warn "Detected golang version doesn't match preferred Go version for Origin."
		os::log::warn "This version mismatch could lead to differences in execution between this run and the Origin CI systems."
		return 0
	fi
fi
```

stevekuznetsov replied:
Specifically it is very important that if someone is using this flag they are aware that there is a very real chance that everything could look great locally but the stuff they see on the CI will not be the same.

soltysh replied:
Makes sense, updated.

soltysh commented:
@stevekuznetsov comments addressed.

stevekuznetsov commented:
Bash LGTM

soltysh commented:
```
===== Verifying UPSTREAM Commits =====

SUCCESS: All commits are valid.

hack/verify-gofmt.sh

[INFO] Detected go version: go version devel +111d590 Fri Jul 29 01:09:55 2016 +0000 linux/amd64.

[ERROR] Please install Go version 1.6 or use PERMISSIVE_GO=y to bypass this check.
```
@stevekuznetsov what's your plan to deal with the tip version of go we use in travis?

stevekuznetsov commented:
I'd have to ask @smarterclayton what, exactly, we gain by running a tip build on Travis first. We could just pass the permissive flag in to Travis.

smarterclayton commented:
We do it to flag early go build problems coming from upstream.

stevekuznetsov commented:
Ok, then letting it be permissive should be fine, we can catch updates for
`go tool vet` as well.

On Jul 30, 2016 6:30 PM, "Clayton Coleman" <notifications@github.com> wrote:

> We do it to flag early go build problems coming from upstream.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/10101#issuecomment-236393390>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AG_ScjvRvfbNm0jbu_ylRbJ_rCpCCLmBks5qa9BmgaJpZM4JYBXs>
> .
>

soltysh commented:
@stevekuznetsov for travis part and `pkg/auth/ldaputil/client.go` since you touched it
@liggitt for the remaining go code

soltysh commented:
[test]

stevekuznetsov commented:
re[test]

stevekuznetsov commented:
re[test]

soltysh commented:
@stevekuznetsov I've modified the go check once again, since it wasn't catching go1.6.x properly
@liggitt I'm removing the `pkg/auth/ldaputil/client.go` modifications for now, I've added there only a TODO and we will see the build for tip go fail on travis each time.

liggitt commented:
-1 on making Travis always go red

soltysh commented:
Apparently go has a [clone](https://github.com/golang/go/commit/db5802104797cadcb4f44c5198a0fc39e13f9bc3) method for that :/

liggitt commented:
Yuck

soltysh commented:
Due to https://github.com/golang/go/issues/15771 I've created our local clone method for `tls.Config` with appropriate FIXME pointing to that issue. @liggitt ptal

soltysh commented:
```
pkg/auth/ldaputil/client.go:145: unknown tls.Config field 'DynamicRecordSizingDisabled' in struct literal
pkg/auth/ldaputil/client.go:146: unknown tls.Config field 'Renegotiation' in struct literal
```

and of course they don't exists in go 1.6, grrr....

stevekuznetsov commented:
@soltysh it is unfortunate they chose not to export this ... at least go1.8 is only a year away :*(

stevekuznetsov commented:
@openshift-bot, the [last build](https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/7550/) failed from the following flakes:
 - [check](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin_check/4626/consoleFull#-212491224457692177e4b031e1b25cfb77) failed on: https://github.com/openshift/origin/issues/9457

re[test]

soltysh commented:
@liggitt, Steve suggested we go with ignoring certain errors and that's the final solution. @stevekuznetsov ptal

stevekuznetsov commented:
> @liggitt, Steve suggested we go with ignoring certain errors and that's the final solution

To expand on this, the reason was that there is no sane way to have a `Copy()` method for the `TLSConfig` that satisfied the 1.6 and 1.7 golang stdlibs, so we will continue to do the copy unsafely and just follow up by placing a new lock into the copy manually after the fact (@soltysh that's still a TODO)

stevekuznetsov commented on hack/verify-govet.sh:
```diff
@@ -3,82 +3,49 @@
+	"pkg/auth/ldaputil/client.go:[[:digit:]]+: assignment copies lock value to c: crypto/tls.Config contains sync.Once contains sync.Mutex"
+)
 
-DIR_BLACKLIST='./hack
```
@soltysh we don't want to remove this part of `hack/verify-govet.sh`

soltysh replied:
Why, I haven't see us using it?

stevekuznetsov commented on hack/verify-govet.sh:
```diff
@@ -3,82 +3,49 @@
-# all top-level directories except for 'pkg', and all second-level subdirectories of 'pkg'.
-ALL_DIRS=$(find_files | grep -Eo "\./([^/]+|pkg/[^/]+)" | sort -u)
+vetexceptions=(
+	"pkg/auth/ldaputil/client.go:[[:digit:]]+: assignment copies lock value to c: crypto/tls.Config contains sync.Once contains sync.Mutex"
```
I'm not sure MacOS will support `[[:digit:]]` -- why not `\d` or `[0-9]`?

soltysh replied:
`\d` doesn't work, `[0-9]` seems reasonable, will go with that.

stevekuznetsov commented on hack/verify-govet.sh:
```diff
@@ -3,82 +3,49 @@
+function is_vetexception() {
+	text=$1
+	for ve in "${vetexceptions[@]}"; do
+		if [[ -n $(echo "${text}" | grep -Eox "${ve}") ]]; then
```
just check the return value
```sh
if grep -Eqx "${ve}" <<<"${text}"; then
    # the text we got matches this blacklist entry
    return 0
fi
```

stevekuznetsov commented on hack/verify-govet.sh:
```diff
@@ -3,82 +3,49 @@
-./third_party
-./tools'
+function is_vetexception() {
+	text=$1
```
This should be `local`

stevekuznetsov commented on hack/verify-govet.sh:
```diff
@@ -3,82 +3,49 @@
-./test
-./third_party
-./tools'
+function is_vetexception() {
```
I would call this `govet_blacklist_contains`

stevekuznetsov commented on hack/verify-govet.sh:
```diff
@@ -3,82 +3,49 @@
-# directories we're searching through. The following are all of the directories we care about:
-# all top-level directories except for 'pkg', and all second-level subdirectories of 'pkg'.
-ALL_DIRS=$(find_files | grep -Eo "\./([^/]+|pkg/[^/]+)" | sort -u)
+vetexceptions=(
```
How about `govet_blacklist`?

stevekuznetsov commented on hack/verify-govet.sh:
```diff
@@ -3,82 +3,49 @@
-./tools'
+function is_vetexception() {
+	text=$1
+	for ve in "${vetexceptions[@]}"; do
```
Naming nit, how do you like
```sh
for blacklist_entry in "${govet_blacklist[@]}"; do
```

stevekuznetsov commented on hack/verify-govet.sh:
```diff
@@ -3,82 +3,49 @@
-    fi
-  fi
+FAILURE=false
+test_dirs=$(find_files | cut -d '/' -f 1-2 | sort -u)
```
Since we're in here lets do some Bash cleanup of my legacy code :)

Quote the expansion:
```sh
test_dirs="$(find_files | cut -d '/' -f 1-2 | sort -u)"
          ^                                           ^
```

stevekuznetsov commented on hack/verify-govet.sh:
```diff
@@ -3,82 +3,49 @@
-  fi
+FAILURE=false
+test_dirs=$(find_files | cut -d '/' -f 1-2 | sort -u)
+for test_dir in $test_dirs; do
```
Since we're in here lets do some Bash cleanup of my legacy code :)

Bracket the var:
```sh
for test_dir in ${test_dirs}; do
                 ^         ^
```

stevekuznetsov commented on hack/verify-govet.sh:
```diff
@@ -3,82 +3,49 @@
+				echo "${result}"
+				FAILURE=true
+			fi
+		done <<<${result}
```
```sh
done <<<"${result}"
        ^         ^
```

stevekuznetsov commented on hack/verify-govet.sh:
```diff
@@ -3,82 +3,49 @@
-then
-  echo "FAILURE: go vet failed!"
-  exit 1
+if $FAILURE; then
```
Since we're in here lets do some Bash cleanup of my legacy code :)

Use proper check in the `if`:

```sh
if [[ -n "${FAILURE:-}" ]]; then
```

stevekuznetsov commented on pkg/cmd/util/net.go:
```diff
@@ -123,8 +124,9 @@ func TransportFor(ca string, certFile string, keyFile string) (http.RoundTripper
 	// Copy default transport
-	transport := *http.DefaultTransport.(*http.Transport)
-	transport.TLSClientConfig = &tls.Config{}
+	transport := knet.SetTransportDefaults(&http.Transport{
```
@soltysh what did we gain from using the k8s lib?

liggitt replied:
the intent of this was to pick up default transport settings, which the lib does safely

soltysh commented:
@stevekuznetsov updated except for the removed chunk, I don't see why we need it if it's of no use, unless I'm missing something?

soltysh commented:
ptal

soltysh commented:
Reverted back the block @stevekuznetsov was asking for and added a two more ignores to the generated clientset, which will require a bit more of work to have them pass govet.

stevekuznetsov commented:
LGTM on a green [test] -- maybe we should open issues for those lock copies that are our fault

soltysh commented:
Issue created https://github.com/openshift/origin/issues/10357.

soltysh commented:
Tests are green.

[merge]
